programs/mango-v4/src/instructions/token_force_close_borrows_with_token.rs

```diff
@@ -0,0 +1,207 @@
+    // Transfer corresponding amount of asset_token from liqee to liqor.
+    //
+    let now_ts = Clock::get()?.unix_timestamp.try_into().unwrap();
+    {
+        let account_retriever: &mut ScanningAccountRetriever = &mut account_retriever;
+        let liab_token_index = liab_token_index;
+        let asset_token_index = asset_token_index;
+        let liqor: &mut MangoAccountRefMut = &mut liqor.borrow_mut();
+        let liqor_key = ctx.accounts.liqor.key();
+        let liqee: &mut MangoAccountRefMut = &mut liqee.borrow_mut();
+        let liqee_key = ctx.accounts.liqee.key();
+
+        let (asset_bank, asset_oracle_price, opt_liab_bank_and_price) =
+            account_retriever.banks_mut_and_oracles(asset_token_index, liab_token_index)?;
+        let (liab_bank, liab_oracle_price) = opt_liab_bank_and_price.unwrap();
+        // account constraint #2
+        require!(liab_bank.is_force_close(), MangoError::SomeError);
+
+        let (liqee_asset_position, liqee_asset_raw_index) =
+            liqee.token_position_and_raw_index(asset_token_index)?;
+        let liqee_asset_native = liqee_asset_position.native(asset_bank);
+        // TODO: should we enforce that asset position is positive? or just incur borrows if needed, I think borrowing is fine
+
+        let (liqee_liab_position, liqee_liab_raw_index) =
+            liqee.token_position_and_raw_index(liab_token_index)?;
+        let liqee_liab_native = liqee_liab_position.native(liab_bank);
+        require!(liqee_liab_native.is_negative(), MangoError::SomeError);
+
+        // The amount of liab native tokens we will transfer
+        let liab_transfer = min(-liqee_liab_native, max_liab_transfer);
+
+        // The amount of asset native tokens we will give up for them
+        let fee_factor = I80F48::ONE + liab_bank.liquidation_fee;
+        let liab_oracle_price_adjusted = liab_oracle_price * fee_factor;
+        let asset_transfer = liab_transfer * liab_oracle_price_adjusted / asset_oracle_price;
+
+        // Apply the balance changes to the liqor and liqee accounts
+        let liqee_liab_position = liqee.token_position_mut_by_raw_index(liqee_liab_raw_index);
+        let liqee_liab_active =
+            liab_bank.deposit_with_dusting(liqee_liab_position, liab_transfer, now_ts)?;
+        let liqee_liab_indexed_position = liqee_liab_position.indexed_position;
+
+        let (liqor_liab_position, liqor_liab_raw_index, _) =
+            liqor.ensure_token_position(liab_token_index)?;
+        let liqor_liab_native = liqor_liab_position.native(liab_bank);
+        // Require that liqor obtain deposits before he tries to liquidate liqee's borrows, to prevent
+        // moving other liquidator from earning further liquidation fee from these borrows by trying to liquidate the current liqor
+        require!(liqor_liab_native.is_positive(), MangoError::SomeError);
+        require_gte!(
+            liqor_liab_native.abs(),
+            liqee_liab_native.abs(),
```

Comment:
I don't like this because the caller may have specified a smaller max_settle -- better to fold this into the liab_transfer computation above